src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1018,6 +1018,16 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     vdef1
   }
 
+  /** Widen type if a valDef is both mutable and final, see #1285 */
+  private def widenIfFinalMutable(vdef: Tree, sym: Symbol)(implicit ctx: Context) = {
+    if (sym.is(allOf(Mutable, Final))) {
+      val widened = vdef.tpe.widen
+      vdef.withType(widened)
+      sym.info_=(widened)
+    }
+    vdef
+  }
+
   /** Add a @volitile to lazy vals when rewriting from Scala2 */
   private def patchIfLazy(vdef: ValDef)(implicit ctx: Context): Unit = {
```

Comment:
There's a simpler way to do it. In Namer.scala

      def isInline = sym.is(Final, butNot = Method)

We just need to exclude Mutable here. E.g.

      def isInline = sym.is(Final, butNot = Method | Mutable)